vscode/src/debugger/session.ts

```diff
@@ -144,57 +190,392 @@ export class QscDebugSession extends LoggingDebugSession {
+      log.trace(`raising breakpoint event`);
+      const evt = new StoppedEvent(
+        "breakpoint",
+        QscDebugSession.threadID
+      ) as DebugProtocol.StoppedEvent;
+      evt.body.hitBreakpointIds = [res];
+      this.sendEvent(evt);
+
+      this.sendEvent(
+        new BreakpointEvent("changed", {
+          verified: true,
+          id: res,
+        } as DebugProtocol.Breakpoint)
+      );
+    } else {
+      this.endSession(`ending session`);
+    }
+  }
+
+  private endSession(message: string) {
+    log.trace(message);
+    vscode.debug.activeDebugConsole.appendLine("");
+    vscode.debug.activeDebugConsole.appendLine(SimulationCompleted);
+    this.sendEvent(new TerminatedEvent());
+    this.sendEvent(new ExitedEvent(0));
+  }
+
+  private async runWithoutDebugging(
+    args: ILaunchRequestArguments
+  ): Promise<void> {
+    const bps: number[] = [];
+    for (let i = 0; i < args.shots; i++) {
+      this.run(bps);
+    }
+  }
+
+  private get_breakpoint_ids(): number[] {
+    const bps: number[] = [];
+    for (const bp of this.breakpoints.get(this.program) ?? []) {
+      if (bp && bp.id) {
+        bps.push(bp.id);
+      }
+    }
+
+    return bps;
+  }
+  protected async continueRequest(
+    response: DebugProtocol.ContinueResponse,
+    args: DebugProtocol.ContinueArguments
+  ): Promise<void> {
+    log.trace(`continueRequest: %O`, args);
+    const bps = this.get_breakpoint_ids();
+
+    log.trace(`sending continue response`);
+    this.sendResponse(response);
+
+    const eventTarget = createDebugConsoleEventTarget();
+    await this.debugService.evalContinue(bps, eventTarget).then(
+      (res) => {
+        if (res) {
+          log.trace(`raising breakpoint event`);
+          this.sendEvent(
+            new StoppedEvent("breakpoint", QscDebugSession.threadID)
+          );
+          this.sendEvent(
+            new BreakpointEvent("changed", {
+              verified: true,
+              id: res,
+            } as DebugProtocol.Breakpoint)
+          );
+        } else {
+          this.endSession(`ending session`);
+        }
+      },
+      (e) => {
+        log.info(`Runtime error: ${e}`);
+        this.endSession(`ending session`);
+      }
+    );
+  }
+
+  protected async breakpointLocationsRequest(
+    response: DebugProtocol.BreakpointLocationsResponse,
+    args: DebugProtocol.BreakpointLocationsArguments,
+    request?: DebugProtocol.Request
+  ): Promise<void> {
+    log.trace(`breakpointLocationsRequest: %O`, args);
+
+    response.body = {
+      breakpoints: [],
+    };
+
+    const fileUri = vscode.Uri.parse(args.source.path ?? "", false);
+
+    const file = vscode.workspace.textDocuments.find(
+      (td) => td.uri.path === fileUri.path
+    );
+    if (!file) {
+      for (const td of vscode.workspace.textDocuments) {
+        log.trace("breakpointLocationsRequest: potential file" + td.uri.path);
+      }
+      log.trace("breakpointLocationsRequest: target file" + fileUri.path);
```

Comment:
Nice! this tracing will be helpful as we continue to wrestle with the virtual file system issues.